core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -857,6 +858,364 @@ KOKKOS_INLINE_FUNCTION_DELETED Impl::ThreadVectorRangeBoundariesStruct<
 ThreadVectorRange(const TeamMemberType&, const iType1& arg_begin,
                   const iType2& arg_end) = delete;
 
+// FIXME just move these up to actual Impl namespace
+namespace Impl {
+enum class MDTeamRangeLastRank : bool { NotLastRank, LastRank };
+enum class MDTeamRangeParThread : bool { NotParThread, ParThread };
+enum class MDTeamRangeParVector : bool { NotParVector, ParVector };
+enum class MDTeamRangeThreadAndVector : bool { NotBoth, Both };
+
+struct NoReductionTag {
+  explicit NoReductionTag() = default;
+};
+
+inline constexpr NoReductionTag no_reduction_tag{};
+
+template <MDTeamRangeLastRank LR, MDTeamRangeParThread PT,
+          MDTeamRangeParVector PV>
+struct MDTeamRangeStatus {
+  static constexpr MDTeamRangeLastRank last_rank   = LR;
+  static constexpr MDTeamRangeParThread par_thread = PT;
+  static constexpr MDTeamRangeParVector par_vector = PV;
+};
+
+template <int CurrentRank, int TotalRank, int ParThreadRank, int ParVectorRank,
+          Iterate Iter>
+struct MDTeamRangeNextStatus {
+  using rank_type                            = int;
+  static constexpr rank_type current_rank    = CurrentRank;
+  static constexpr rank_type total_rank      = TotalRank;
+  static constexpr rank_type par_thread_rank = ParThreadRank;
+  static constexpr rank_type par_vector_rank = ParVectorRank;
+  static constexpr Iterate iter              = Iter;
+
+  // We have to recursively process ranks [0..total_rank-1]
+  using type = MDTeamRangeStatus<
+      (iter == Iterate::Right)
+          ? static_cast<MDTeamRangeLastRank>(current_rank == total_rank)
+          : static_cast<MDTeamRangeLastRank>(current_rank == -1),
+      static_cast<MDTeamRangeParThread>(current_rank == par_thread_rank),
+      static_cast<MDTeamRangeParVector>(current_rank == par_vector_rank)>;
+};
+
+template <int CurrentRank, int TotalRank, int ParThreadRank, int ParVectorRank,
+          Iterate Iter>
+using MDTeamRangeStatus_t =
+    typename MDTeamRangeNextStatus<CurrentRank, TotalRank, ParThreadRank,
+                                   ParVectorRank, Iter>::type;
+
+template <int totalRank, Iterate Iter, typename ExecSpace,
+          MDTeamRangeThreadAndVector t_and_v>
+struct ParallelRank;
+
+// FOR SERIAL ONLY
```

Comment:
?? What does this mean?

Reply:
Removed in [ed4859d](https://github.com/kokkos/kokkos/pull/5238/commits/ed4859d102a0817c046a461751a5587dd5fc2a69) for now.

This was just a placeholder. Initially, we were thinking of adding in structs specialized for the other backends as well, but we'll do this part once the base implementation complete.
I was also exploring some options to define each `ParallelRank` struct for other backends in each file located in their respective backend directories, so that we can avoid keep adding in ifdef'ed `ParallelRank` struct all in here.